notes-0319-inductive-types.tex

```diff
@@ -0,0 +1,228 @@
+
+\begin{prooftree*}
+  \hypo{x : \N : \rightarrow P : U}
+  \hypo{M_0 : P[zero/x]}
+  \hypo{ O : \N}
+  \infer[no rule]3{y : P[x'/x] \rightarrow M_{suc} : P[(suc x' / x]}
+  \infer1[$N$-Elimination]{rec_\N [x.P] (M_0 ; x.y.M_{succ} ; O) : P[O/x]}
+\end{prooftree*}
+
+With traditional induction on natural numbers, induction is a black box, we don't necessarily know what the proofs for each of the $P(n)$ we are proving look like. In type theory, we want to go one step farther now and say exactly what the proof of $P(n)$ look like for each $n$. This motivates two computation rules, that tell us which proof of $P(n)$ (or equivalently which element $p : P(n)$) is constructed by the induction principle for the different values of $O$:
+
+\begin{prooftree*}
+  \hypo{x : \N : \rightarrow P : U}
+  \hypo{M_0 : P[zero/x]}
+  \infer[no rule]2{y : P[x'/x] \rightarrow M_{suc} : P[(suc\ x' / x]}
+  \infer1[$0$-Computation]{rec_{\N} [x.P] (M_0 ; x.y.M_{suc} ; zero) \equiv M_0 : P[zero/x]}
+\end{prooftree*}
+
+and
+
+\begin{prooftree*}
+  \hypo{x : \N : \rightarrow P : U}
+  \hypo{M_0 : P[zero/x]}
+  \hypo{ A : \N}
+  \infer[no rule]3{y : P[x'/x] \rightarrow M_{suc} : P[(suc\ x' / x]}
+  \infer1[$0$-Computation]{rec_{\N} [x.P] (M_0 ; x.y.M_{suc} ; suc\ A)) \equiv M_{suc} [\frac{A, rec_\N [x.P][M_0;x.y.M_{suc};A)}{x , y}] : P[\frac{suc\ A}{x}]}
+\end{prooftree*}
+
+
+The first rule says the elimination rule when applied to $zero$ is exactly the proof $M_0$ of the zero case.
+The second says that the elimination rule applied to a successor $suc\ n$ is exactly the proof $M_{suc}$ applied to a recursive call to the elimination rule for $n$.
+
+\begin{remark}
+  While it was perfectly valid for us to have a constructor of type $\N \rightarrow \N$ when defining an inductive type, not all types make sense as the type of a constructor.
+  For example if we want to define a type $C$, a constructor of type $(C \rightarrow 0) \rightarrow C$ is inconsistent because the corresponding recursion principle would allow us to construct an element of type $0$.
+  Another example of an inconsistent type for a constructor is $((D \rightarrow Prop) \rightarrow Prop) \rightarrow D)$.
+  To keep our types consistent it suffices to restrict ourselves to constructors whose domains have strict positivity.
+\end{remark}
+
+
+\section{Initial Algebras}
+One way to formalize inductive types is to give them a universal property as initial objects in a category of appropriate ``algebras'', where the ``algebras'' in this category are types with the same structure as the structure imposed by the constructors.
+This ``algebra'' structure will ensure that every object in the category is in some sense ``at most'' the inductive type we want to define.
```

Comment:
"at most" could be misleading here because, for example, N * N can still form a N-algebra by ignoring the second component. Otherwise this paragraph looks great!